File: crates/workspaces/src/adapters/app_icons.m
// Application icon lookup, barik-style: resolve the running application by
// bundle identifier (preferred) or localized name, then encode NSWorkspace's
// icon for its bundle as PNG. Exports plain C functions; Rust never touches
// the Obj-C objects.

#import <AppKit/AppKit.h>

// PNG bytes of the icon for the application identified by `bundle_id`
// (preferred) or `app_name`, rendered at `pixel_size`. On success writes a
// malloc'd buffer to *out_bytes — free it with workspaces_free_icon_png — and
// returns its length. Returns 0 when the application or icon cannot be
// resolved.
int64_t workspaces_copy_app_icon_png(const char *bundle_id, const char *app_name,
                                     int32_t pixel_size, uint8_t **out_bytes) {
    @autoreleasepool {
        NSString *bundle = (bundle_id && bundle_id[0])
                               ? [NSString stringWithUTF8String:bundle_id]
                               : nil;
        NSString *name = (app_name && app_name[0])
                             ? [NSString stringWithUTF8String:app_name]
                             : nil;

        NSRunningApplication *app = nil;
        if (bundle) {
            app = [NSRunningApplication runningApplicationsWithBundleIdentifier:bundle]
                      .firstObject;
        }
        if (app == nil && name) {
            for (NSRunningApplication *candidate in
                 [NSWorkspace sharedWorkspace].runningApplications) {
                if ([candidate.localizedName isEqualToString:name]) {
                    app = candidate;
                    break;
                }
            }
        }

        NSImage *icon = nil;
        if (app.bundleURL != nil) {
            icon = [[NSWorkspace sharedWorkspace] iconForFile:app.bundleURL.path];
        }
        if (icon == nil && bundle) {
            NSURL *url = [[NSWorkspace sharedWorkspace]
                URLForApplicationWithBundleIdentifier:bundle];
            if (url != nil) {
                icon = [[NSWorkspace sharedWorkspace] iconForFile:url.path];
            }
        }
        if (icon == nil) {
            return 0;
        }

        NSRect rect = NSMakeRect(0, 0, pixel_size, pixel_size);
        CGImageRef cg = [icon CGImageForProposedRect:&rect
                                             context:nil
                                               hints:nil];
        if (cg == NULL) {
            return 0;
        }
        NSBitmapImageRep *rep = [[NSBitmapImageRep alloc] initWithCGImage:cg];
        NSData *png = [rep representationUsingType:NSBitmapImageFileTypePNG
                                        properties:@{}];
        if (png == nil || png.length == 0) {
            return 0;
        }

        uint8_t *buffer = malloc(png.length);
        if (buffer == NULL) {
            return 0;
        }
        memcpy(buffer, png.bytes, png.length);
        *out_bytes = buffer;
        return (int64_t)png.length;
    }
}

void workspaces_free_icon_png(uint8_t *bytes) {
    free(bytes);
}

File: crates/workspaces/src/adapters/app_icons.swift
import AppKit
import Foundation

@_cdecl("workspaces_copy_app_icon_png")
public func workspacesCopyAppIconPNG(
    _ bundleID: UnsafePointer<CChar>?,
    _ appName: UnsafePointer<CChar>?,
    _ pixelSize: Int32,
    _ outBytes: UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>?
) -> Int64 {
    let bundle = bundleID.flatMap { $0.pointee == 0 ? nil : String(cString: $0) }
    let name = appName.flatMap { $0.pointee == 0 ? nil : String(cString: $0) }
    var app = bundle.flatMap { NSRunningApplication.runningApplications(withBundleIdentifier: $0).first }
    if app == nil, let name {
        app = NSWorkspace.shared.runningApplications.first { $0.localizedName == name }
    }
    var icon = app?.bundleURL.map { NSWorkspace.shared.icon(forFile: $0.path) }
    if icon == nil, let bundle, let url = NSWorkspace.shared.urlForApplication(withBundleIdentifier: bundle) {
        icon = NSWorkspace.shared.icon(forFile: url.path)
    }
    guard let icon else { return 0 }
    var rect = NSRect(x: 0, y: 0, width: Int(pixelSize), height: Int(pixelSize))
    guard let image = icon.cgImage(forProposedRect: &rect, context: nil, hints: nil),
          let png = NSBitmapImageRep(cgImage: image).representation(using: .png, properties: [:])
    else { return 0 }
    let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: png.count)
    png.copyBytes(to: buffer, count: png.count)
    outBytes?.pointee = buffer
    return Int64(png.count)
}

@_cdecl("workspaces_free_icon_png")
public func workspacesFreeIconPNG(_ bytes: UnsafeMutablePointer<UInt8>?) {
    bytes?.deallocate()
}
